api/src/game_state_server.py

```diff
@@ -156,93 +104,118 @@ async def handle_connection(
         session_id = str(uuid4())
         with timber.context(connection={'session_id': session_id, 'room_id': room_id}):
             async with self._rate_limiter.rate_limited_connection(client_ip, room_id):
-                room_changes = await self.room_store.changes(room_id)
+                listener_q: asyncio.Queue[Response] = asyncio.Queue()
 
                 with self.apm_transaction('connect'):
-                    result = await self.room_store.apply_mutation(
-                        room_id,
-                        request_id=None,
-                        mutate=lambda entities: self._acquire_room_slot(
-                            room_id, entities, client_ip
-                        ),
+                    if not self._room_context_by_id.get(room_id):
+                        await self._acquire_room_slot(room_id, client_ip)
+                        room = create_room(room_id, await self.room_store.read(room_id))
+                        change_listener = ChangeListener(
+                            [],
+                            asyncio.create_task(
+                                self._listen_for_changes(room_id, session_id)
+                            ),
+                        )
+                        self._room_context_by_id[room_id] = RoomContext(
+                            room, change_listener
+                        )
+                    self._room_context_by_id[
+                        room_id
+                    ].change_listener.output_queues.append(listener_q)
+
+                    room_state = list(
+                        self._room_context_by_id[room_id].room.game_state.values()
                     )
-                    yield Response('connected', result.entities)
+                    yield ConnectionResponse(room_state)
 
-                response_msgs = self._requests_to_messages(room_id, requests)
-                change_msgs = _room_changes_to_messages(room_changes)
-
-                async for msg in amerge(
-                    response_msgs,
-                    change_msgs,
-                    complete_when=CompleteCondition.FIRST_COMPLETED,
-                ):
-                    yield msg
-
-    async def _requests_to_messages(
-        self,
-        room_id: str,
-        requests: AsyncIterator[DecoratedRequest],
-    ) -> AsyncIterator[Response]:
-        async for request in requests:
-            with timber.context(request={'request_id': request.request_id}):
-                async for message in self._handle_request(
-                    updates=request.updates,
-                    room_id=room_id,
-                    request_id=request.request_id,
-                ):
-                    yield message
-
-    async def _handle_request(
+                self._room_occupancy[room_id] += 1
```

Comment:
I think room_occupancy is the same as the number of queues in the change listener's `output_queues`

Reply:
Works for me.